drivers/d3d12/rendering_device_driver_d3d12.cpp

```diff
@@ -2143,17 +2143,27 @@ void RenderingDeviceDriverD3D12::command_pipeline_barrier(CommandBufferID p_cmd_
 	for (uint32_t i = 0; i < p_texture_barriers.size(); i++) {
 		const TextureBarrier &texture_barrier_rd = p_texture_barriers[i];
 		const TextureInfo *texture_info = (const TextureInfo *)(texture_barrier_rd.texture.id);
+		if (texture_info->main_texture) {
+			texture_info = texture_info->main_texture;
+		}
 		_rd_stages_and_access_to_d3d12(p_src_stages, texture_barrier_rd.prev_layout, texture_barrier_rd.src_access, texture_barrier_d3d12.SyncBefore, texture_barrier_d3d12.AccessBefore);
 		_rd_stages_and_access_to_d3d12(p_dst_stages, texture_barrier_rd.next_layout, texture_barrier_rd.dst_access, texture_barrier_d3d12.SyncAfter, texture_barrier_d3d12.AccessAfter);
 		texture_barrier_d3d12.LayoutBefore = _rd_texture_layout_to_d3d12_barrier_layout(texture_barrier_rd.prev_layout);
 		texture_barrier_d3d12.LayoutAfter = _rd_texture_layout_to_d3d12_barrier_layout(texture_barrier_rd.next_layout);
 		texture_barrier_d3d12.pResource = texture_info->resource;
-		texture_barrier_d3d12.Subresources.IndexOrFirstMipLevel = texture_barrier_rd.subresources.base_mipmap;
-		texture_barrier_d3d12.Subresources.NumMipLevels = texture_barrier_rd.subresources.mipmap_count;
-		texture_barrier_d3d12.Subresources.FirstArraySlice = texture_barrier_rd.subresources.base_layer;
-		texture_barrier_d3d12.Subresources.NumArraySlices = texture_barrier_rd.subresources.layer_count;
-		texture_barrier_d3d12.Subresources.FirstPlane = _compute_plane_slice(texture_info->format, texture_barrier_rd.subresources.aspect);
-		texture_barrier_d3d12.Subresources.NumPlanes = format_get_plane_count(texture_info->format);
+		if (texture_barrier_rd.subresources.mipmap_count == texture_info->mipmaps && texture_barrier_rd.subresources.layer_count == texture_info->layers) {
+			// So, all resources. Then, let's be explicit about it so D3D12 doesn't think
+			// we are dealing with a subset of subresources.
+			texture_barrier_d3d12.Subresources.IndexOrFirstMipLevel = 0xffffffff;
+			texture_barrier_d3d12.Subresources.NumMipLevels = 0;
```

Comment:
The for loop doesn't explicitly initialize the values for these attributes, as the structure is reused in every iteration.

The following must also be set on this case:
```cpp
D3D12_TEXTURE_BARRIER::Subresources.FirstArraySlice
D3D12_TEXTURE_BARRIER::Subresources.NumArraySlices
D3D12_TEXTURE_BARRIER::Subresources.FirstPlane
D3D12_TEXTURE_BARRIER::Subresources.NumPlanes
```

Otherwise the behavior of the code will depend on what the previous barrier has specified or the value from the default initialization if it's the first element.

Reply:
As long as `NumMipLevels == 0`, they ignore all those four fields (per https://learn.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_barrier_subresource_range).

Reply:
To counter my own comment: https://learn.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_barrier_subresource_range
```
Index of first array slice in the range. Ignored if NumMipLevels is zero.
Number of array slices in the range. Ignored if NumMipLevels is zero.
First plane slice in the range. Ignored if NumMipLevels is zero.
Number of plane slices in the range. Ignored if NumMipLevels is zero.
```
So this behavior is known to work correctly, but I think a comment indicating as such would be appreciated as a review of it without being aware of that would help.

Reply:
I'll add a comment about it.